client_mdb/serverusage_client_mdb.8
.\" Manpage for ServerUsage-Client-MDB.
.TH serverusage_client_mdb 8 "2012-08-29"
.SH NAME
serverusage_client_mdb \- ServerUsage-Client-MDB control interface
.SH SYNOPSIS
serverusage_client_mdb is a SysV init script that takes one-word commands, defined below:
.PP
.B serverusage_client_mdb
.I command
.PP
.SH DESCRIPTION
.B serverusage_client_mdb
is a front end to the ServerUsage-Client_MDB program.
The ServerUsage-Client program is composed by two main sections: the
serverusage_client_mdb.bin to collect usage statistics from the database and the
serverusage_tcpsender_mdb.bin to send the output to a log server via TCP.
.PP
.B NOTE:
The configuration file is located at
.B /etc/serverusage_client_mdb.conf
.PP
.SH OPTIONS
The \fIcommand\fP can be any one of the following options:
.TP 14
.BI start
Start ServerUsage-Client-MDB.
.TP
.BI stop
Stops ServerUsage-Client-MDB.
.TP
.BI restart
Restarts ServerUsage-Client-MDB.
If the program is not running, it is started.
.BI status
Displays a brief status report.
.BI help
Displays a short help message.
.SH AUTHOR
Nicola Asuni (nicola@fubra.com)
.SH COPYRIGHT
(c) 2012-2012 Fubra Limited
Manor Coach House, Church Hill, Aldershot, Hampshire, GU12 4RQ, UK (http://www.fubra.com - support@fubra.com)
.


CHANGELOG.TXT
6.3.1 (2012-09-03)
	- String to number conversion functions were fixed.

6.3.0 (2012-08-29)
	- Debug mode on server was improved to display milliseconds.

6.2.0 (2012-08-21)
	- RPM spec files were changed to avoid overriding the configuration and database files.
	- Init scripts were improved.

6.1.0 (2012-08-16)
	- This versin contains a new module to collect user statistics from MariaDB database.

6.0.0 (2012-08-09)
	- Dual stack support for both IPv4 and IPv6 was added.

5.3.0 (2012-08-08)
	- Documentation was updated.

5.2.0 (2012-07-18)
	- Switch to using contention-free aggregates for the systemtap script.
	- SystemTap skipped probes are now ignored to avoid early program termination.

5.1.0 (2012-07-17)
	- A segmentation fault problems were fixed.

5.0.0 (2012-07-16)
	- Various memory allocation and segmentation fault problems were fixed.

4.9.0 (2012-07-13)
	- A memory allocation problem was fixed.

4.8.0 (2012-06-13)
	- TCP sender was improved to avoid infinite loops.

4.7.0 (2012-06-13)
	- TCP sender and receiver are set to use IPv6 protocol by default.

4.6.0 (2012-06-08)
	- TCP sender and receiver are now able to reuse previously unclosed connections.

4.5.0 (2012-05-22)
	- The program serverusage_tcpreceiver.c was improved to be thread safe.

4.4.0 (2012-05-21)
	- SQLite transactions where fixed to properly lock/unlock the table.

4.3.0 (2012-05-18)
	- serverusage_tcpreceiver.c was fixed.

4.2.0 (2012-05-18)
	- Bug related to nested transactions were fixed.

4.1.0 (2012-05-18)
	- C scripts were analysed and optimised using a memory profiler.

4.0.0 (2012-05-17)
	- This version uses SQLite as back-end database for the server module.
	- Database is now created by makefile script (and preserved if already exist).
	- Init script is now able to install/uninstall the cron job for the database aggregator.
	- The output of the SystemTap module was changed to be TAB-separated instead of comma-separated.

--------------------------------------------------------------------------------
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
----------------------------------------------------------------------------

3.0.4 (2012-05-11)
	- serverusage_tcpsender is now able to store log data on a local filesystem in case of network errors.
	- PATH_TEMPLOG was added into serverusage_client.conf to define the path of local caching.

3.0.3 (2012-05-08)
	- A memory leak was fixed.

3.0.2 (2012-05-08)
	- Fixed minor syntax errors.

3.0.1 (2012-05-08)
	- Minor documentation improvements.

3.0.0 (2012-05-04)
	- This version introduce multithreading support for the TCP receiver.
	- Documentation was fixed.

--------------------------------------------------------------------------------
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
--------------------------------------------------------------------------------

2.0.9 (2012-05-02)
	- The TCP sender and receiver were modified to not exit on errors.

2.0.8 (2012-05-01)
	- Installation instructions were added on README.markdown
	- SysV Init scripts were improved to be compatible with chkconfig.
	- HTML file example was fixed.

2.0.7 (2012-04-30)
	- A bug on client starting script was fixed.

2.0.6 (2012-04-27)
	- Instructions to create RPMs were added on README.markdown file
	- VERSION file was added.

2.0.5 (2012-04-26)
	- Util files were moved inside the server package.
	- PHP util script was improved to read main configuration file.

2.0.4 (2012-04-25)
	- Makefiles now includes the DESTDIR variable support.
	- SPEC files were added to create RPM packages.

// -----------------------------------------------------------------------------

2.0.003 (2012-04-17)
	- All files were renamed for compatibility issues.

2.0.000 (2012-04-17)
	- Project were splitted in client, server and util sub-projects.
	- All files were renamed.
	- SysV init files were added to start/stop the services.
	- makefile was added to client and server to compile and install the applications.
	- Manpage and README files were added.

--------------------------------------------------------------------------------
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
--------------------------------------------------------------------------------

1.3.015 (2012-04-11)
	- IO probe on serverusage.stp was fixed.

1.3.014 (2012-04-11)
	- SVG output on srvusgapi.php was improved.
	- Mode 'csv' was added to srvusgapi.php to return data as CSV tab-separated values.
	- Mode 'psa' was added to srvusgapi.php to return data as Base64 encoded PHP Serialized Array.

1.3.013 (2012-04-05)
	- The API srvusgapi.php now includes several new parameters and features (check the documentation).
	- Default displaysvg.html now automatically set the graphic size to fit all available space and by default uses the logarithmic scale.

1.3.012 (2012-04-04)
	- SQL error on serverusageagg.sql was fixed.

1.3.011 (2012-04-04)
	- serverusageagg.sql script was changed to exclude swapper process (UID = 0).

1.3.010 (2012-03-29)
	- SQL queries were improved.
	- srvusgapi.php was improved.

1.3.009 (2012-03-27)
	- An example HTML page (displaysvg.html) to display and update the SVG graph was added.
	- SVG content type was fixed.

1.3.008 (2012-03-27)
	- SVG mode on srvusgapi.php were fixed and improved.

1.3.007 (2012-03-26)
	- The queries on srvusgapi.php file were fixed.
	- The parameter 'mode' was added to srvusgapi.php with the ability to output data in SVG mode.

1.3.006 (2012-03-13)
	- Documentation was fixed.
	- Source code was cleaned-up.

1.3.005 (2012-03-13)
	- Database table views were removed.
	- Web service srvusgapi.php was fixed and improved to include new metrics.

1.3.004 (2012-03-12)
	- Aggregation table name were changed.
	- A table view was added to extract some information.
	- Web service srvusgapi.php was added to get data from the view table.

1.3.003 (2012-03-09)
	- The README was updated.


client_mdb/serverusage_client_mdb.c
/*
//=============================================================================+
// File name   : serverusage_client_mdb.c
// Begin       : 2012-08-14
// Last Update : 2012-09-03
// Version     : 6.3.1
//
// Website     : https://github.com/fubralimited/ServerUsage
//
// Description : This program gets the users' usage statistics from MariaDB
//               database and send them to ServerUsage-Server.
//
// Author: Nicola Asuni
//
// (c) Copyright:
//               Fubra Limited
//               Manor Coach House
//               Church Hill
//               Aldershot
//               Hampshire
//               GU12 4RQ
//				 UK
//               http://www.fubra.com
//               support@fubra.com
//
// License:
//		Copyright (C) 2012-2012 Fubra Limited
//
//		This program is free software; you can redistribute it and/or modify
// 		it under the terms of the GNU General Public License as published by
// 		the Free Software Foundation; version 2 of the License.
//
// 		This program is distributed in the hope that it will be useful,
// 		but WITHOUT ANY WARRANTY; without even the implied warranty of
// 		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// 		GNU General Public License for more details.
//
// 		You should have received a copy of the GNU General Public License along
// 		with this program; if not, write to the Free Software Foundation, Inc.,
// 		51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//
// 		See LICENSE.TXT file for more information.
//=============================================================================+
*/

// TO COMPILE (requires MariaDB-devel package):
// gcc -O3 $(mysql_config --cflags) -o serverusage_client_mdb.bin serverusage_client_mdb.c $(mysql_config --libs)

// USAGE EXAMPLES:
// ./serverusage_client_mdb.bin LOGSERVER_IP LOGSERVER_PORT DATABASE_HOST DATABASE_PORT DATABASE_USER DATABASE_PASSWORD SAMPLING_TIME SERVER_ID PATH_TCPSENDER PATH_TEMPLOG
// ./serverusage_client_mdb.bin 127.0.0.1 9930 localhost 3306 "root" "PASSWORD" 60 "host001" "/usr/bin/serverusage_tcpsender_mdb.bin" "/var/log/serverusage_cache.log"

#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>
#include <mysql.h>

/**
 * Macro to find the max of two numbers
 */
#ifndef max
	#define max(a,b) (((a)>(b))?(a):(b))
#endif

/**
 * String buffer length.
 */
#define BUFLEN 512

/**
 * Report error message and close the program.
 */
void diep(const char *s) {
	perror(s);
	exit(1);
}

/**
 * Main function.
 * @param argc Argument counter.
 * @param argv[] Array of arguments: port, max_connections, mysql_server, mysql_user, mysql_password, mysql_database, mysql_table.
 */
int main(int argc, char *argv[]) {

	// decode arguments
	if (argc != 11) {
		diep("This program collects users statistics for MariaDB database and send them to Serverusage-Server\n\
		You must provide 10 arguments: LOGSERVER_IP LOGSERVER_PORT DATABASE_HOST DATABASE_PORT DATABASE_USER DATABASE_PASSWORD SAMPLING_TIME SERVER_ID PATH_TCPSENDER PATH_TEMPLOG\n\
		FOR EXAMPLE:\n\
		./serverusage_client_mdb.bin 127.0.0.1 9930 \"localhost\" 3306 \"root\" \"PASSWORD\" 60 \"host001\" \"/usr/bin/serverusage_tcpsender_mdb.bin\" \"/var/log/serverusage_cache.log\"");
	}

	// set input values

	// IP of machine running ServerUsage-Server
	char *serverip = (char *)argv[1];
	// TCP port of machine running ServerUsage-Server
	char *serverport = (char *)argv[2];
	// MariaDB hostname
	char *dbserver = (char *)argv[3];
	// MariaDB port
	int dbport = atoi(argv[4]);
	// MariaDB user
	char *dbuser = (char *)argv[5];
	// MariaDB password
	char *dbpassword = (char *)argv[6];
	// sampling time in seconds
	int smp = atoi(argv[7]);
	// Server ID
	char *srvid = (char *)argv[8];
	// Full path to serverusage_tcpsender_mdb.bin
	char *tcpsender = (char *)argv[9];
	// Full path to temporary cache file
	char *cachefile = (char *)argv[10];


	// set command to send data to ServerUsage-Server
	char tcpcmd[BUFLEN];
	// clear the string
	memset(tcpcmd, 0, BUFLEN);
	// compose the command
	sprintf(tcpcmd, "%s \"%s\" \"%s\" \"%s\"", tcpsender, serverip, serverport, cachefile);

	// database query used to get statistics
	char statquery[] = "SELECT USER, CPU_TIME, BYTES_SENT, BYTES_RECEIVED FROM information_schema.USER_STATISTICS";

	// MySQL variables
	MYSQL *dbconn;
	MYSQL_RES *dbresult;
	MYSQL_ROW dbrow;
	int db_num_fields;

	// initialize database conenction
	dbconn = mysql_init(NULL);

	// connect to database
	if (!mysql_real_connect(dbconn, dbserver, dbuser, dbpassword, NULL, dbport, NULL, 0)) {
		diep(mysql_error(dbconn));
	}

	// get CPU ticks per second
    long tps = sysconf(_SC_CLK_TCK);

	time_t starttime;
	time_t readtime;
	double timediff;

	// get starting time;
	starttime = time(NULL);

	// file pointer used to pipe output to serverusage_tcpsender_mdb.bin
	FILE *fp;

	while(1) {
		// get current time in seconds since EPOCH
		readtime = time(NULL);
		// time elapsed since last output
		timediff = difftime(readtime, starttime);
		// print data only at specific timestamps (to keep sampling window synchronized with scripts running on different servers)
		if ((timediff > 0) && (((long)readtime % smp) == 0)) {
			// check for the correct sampling interval
			if (timediff == smp) {
				// execute query to get stats
				if (!(mysql_query(dbconn, statquery))) {
					if ((dbresult = mysql_store_result(dbconn)) != NULL) {
						if ((db_num_fields = mysql_num_fields(dbresult)) == 4) {
							// open the pipe
							if ((fp = popen(tcpcmd, "w")) != NULL) {
								// mark the start of this block of data with an empty line
								fputs("@S\t0\t0\t0\t0\t0\t0\t0\t0\t0\t0\t0\n", fp);
								while (dbrow = mysql_fetch_row(dbresult)) {
									if (dbrow[0] != NULL) {
										// NOTE: the UID is set to 0 and the user name is set to the IP field
										fprintf(fp, "@@\t%lu\t%lu\tmysql\t%s\t0\t%s\t%lu\t%lu\t%lu\t0\t0\n", (long)starttime, (long)readtime, srvid, dbrow[0], max(0, (long)(atof(dbrow[1]) * tps)), max(0, atol(dbrow[2])), max(0, atol(dbrow[3])));
									}
								}
								// mark the end of this block of data with an empty line
								fputs("@E\t0\t0\t0\t0\t0\t0\t0\t0\t0\t0\t0\n", fp);
								// close the pipe
								pclose(fp);
							}
						}
						// clear database results
						mysql_free_result(dbresult);
					}
				}
			}
			// flush stats table
			if (mysql_query(dbconn, "FLUSH USER_STATISTICS")) {
				diep(mysql_error(dbconn));
			}
			// update starting sampling time
			starttime = readtime;
			readtime = time(NULL);
			timediff = difftime(readtime, starttime);
			// wait
			sleep(smp - timediff - 1);
		}
	}

	// close database connection
	mysql_close(dbconn);

	// close program and return 0
	return 0;
}

//=============================================================================+
// END OF FILE
//=============================================================================+
